### ISSUE TITLE
Allow cancelling of changeset uploading

### ISSUE THREAD

**Issue by hikemaniac** [openstreetmap/iD#4717]
Today I made the mistake to upload a changeset with bad internet connection. In the middle of the uploading progress, the connection collapsed and the uploading progress spinner kept spinning for quite a while. I realised what had happened, but of course there was no way to cancel uploading. So I was forced to close the tab and when I reopened it with better connection all mapping was lost. iD couldn't restore my changes, because another tab with changes was open in the background.

Maybe we could implement a way to cancel changeset uploading. JOSM also has such a feature. Would it be possible to add something like this to iD as well?

**Comment by bhousel** [comment]
We could `abort()` the changeset upload POST, but I'm worried that the changes might actually get saved anyway, e.g. if the server has received the request and is just taking a while to process it.   I don't want to rule it out, but we definitely have to think it through.

>  iD couldn't restore my changes, because another tab with changes was open in the background.

By the way, if this happens, you should just be able to close all but one tab, then refresh and get your changes back.

**Comment by manfredbrandl** [comment]
> By the way, if this happens, you should just be able to close all but one tab, then refresh and get your changes back.
@bhousel Do you agree that there should be a common place (http://ideditor.com e.g.) to get this and many other helpfully hints? I would like to see extended documentation somewhere.

**Comment by mmd-osm** [comment]
> We could abort() the changeset upload POST, but I'm worried that the changes might actually get saved anyway, e.g. if the server has received the request and is just taking a while to process it.

That's at least the case for JOSM in my tests. Even if you cancel the upload, the server keeps on processing the OsmChange message. This might take a few minutes in the worst case (btw: upload redesign is underway to significantly speed up the whole backend processing). With a decent internet connection, the upload itself is usually quite fast.

**Comment by bhousel** [comment]
> @bhousel Do you agree that there should be a common place (http://ideditor.com e.g.) to get this and many other helpfully hints? I would like to see extended documentation somewhere.

I think we already have a lot of places to look for information:
- built into iD Help pane
- https://wiki.openstreetmap.org  (a mediawiki)
- https://help.openstreetmap.org  (a stack exchange)
- https://forum.openstreetmap.org  (a phpbb-style forum)
- https://www.reddit.com/r/openstreetmap/  (a subreddit)
- http://learnosm.org/en/  (a github-hosted jekyll site)

I agree that it would be useful to create a more extended documentation for people who really want to improve their editing - but I'm really wary about creating yet another site.  Maybe we can just add extended documentation to the OSM wiki?

**Comment by manfredbrandl** [comment]
> Maybe we can just add extended documentation to the OSM wiki?
The OSM wiki contains nearly no information about iD, it links to a tutorial at http://learnosm.org/en/beginner/id-editor/ which is also outdated. The most up to date information about iD is contained inside iD. I suggest that we add a section "Pro Tips" to the help. @bhousel would you agree?

**Comment by mmd-osm** [comment]
The changeset upload redesign I mentioned in January is now available for testing, see my blog post for details: https://www.openstreetmap.org/user/mmd/diary/44318

**Comment by matkoniecz** [comment]
maybe it can be closed as fixed upstream

if it is not fixed then it requires changes upstream to cleanly revert partially uploaded or just uploaded changeset - is such request open or already rejected?

**Comment by bhousel** [comment]
Seems like it died years ago:  https://github.com/openstreetmap/openstreetmap-website/issues/2201#issuecomment-663608388

Probably best to work under the assumption that changesets inflight can't be cancelled.

**Comment by mmd-osm** [comment]
There's really no such thing as "partially uploaded" (ignoring what JOSM does with chunked uploads, which is a different story). Either the diff applied completely or not at all. These days, upload processing on the server takes around 50-100ms on average. In case you're on an extremely slow network, you might manage to abort the network communication to the server, but that's about it. Once the diff has reached the server, trying to cancel anything in the browser is kind of futile.

Idempotency is relevant, if there's no response from the server for whatever reason, and you want to avoid creating duplicate data when resending the exact same data. Since it's not implemented, diffs with "create new objects"-only might create duplicates in that case. Any diff with modify or delete operations will fail on repeated upload attempts due to version conflicts.

TL;DR: i would just close this bug as not actionable.
